### DISCUSSION TITLE
[Django] This Serverless Function has crashed.

### DISCUSSION THREAD

**Discussion by aikosulaiman** [vercel/community#2292] [Help]
### Summary

I have a django web that runs without no problem in local, but when i tried to deploy it in vercel it says This Serverless Function has crashed.Your connection is working correctly. Vercel is working correctly. 500: INTERNAL_SERVER_ERROR
<img width="954" alt="image" src="https://user-images.githubusercontent.com/70503924/236116495-aae4474b-dd0c-4a19-b451-ca0429442a43.png">
I tried to check the logs and all of the errors are like this:
<img width="962" alt="image" src="https://user-images.githubusercontent.com/70503924/236116626-86c2efea-8f3a-4be2-a85e-19dddd596551.png">

I still have no idea what caused the error, can anybody help me?

### Additional information

My wsgi.py is like this:
<img width="750" alt="image" src="https://user-images.githubusercontent.com/70503924/236116750-8a3e5bb0-2ca7-42a9-bf9d-1e22f6ff5df7.png">
And in the error logs it says Unknown application error occurred
Runtime.Unknown

**Comment by martinekpavel** [comment]
I have same error

**Comment by Gunarjith** [comment]
same error

**Comment by mungruez** [comment]
![image](https://github.com/vercel/community/assets/16914862/02dcc178-17f4-4e3f-92ff-a484d14f7cb6)


Me three !!! Same Error!

**Comment by joshuangdev** [comment]
Same error. I'm using flask though #2779

**Comment by snexuz** [comment]
Same here, is there any alternative django service?

Here's my vercel.json
![image](https://github.com/vercel/community/assets/5488329/23da3ef4-1c6b-4d8b-97c1-b8b51cdf8a67)

**Comment by adinortey360** [comment]
Same error here. I even tried bugsnag to try and catch the error with no success.

**Comment by Abdul12majid** [comment]
Still no solution to this issue ? I just encountered same.

**Reply by adinortey360** [reply]
Are you using a virtual environment?

**Reply by Deepak1649** [reply]
yes I'm and getting the same error

**Comment by andre-fajar-n** [comment]
is there still no solution to this issue?

**Comment by Tusar0003** [comment]
same error here

**Comment by mungruez** [comment]
> ### Summary
> I have a django web that runs without no problem in local, but when i tried to deploy it in vercel it says This Serverless Function has crashed.Your connection is working correctly. Vercel is working correctly. 500: INTERNAL_SERVER_ERROR <img alt="image" width="954" src="https://user-images.githubusercontent.com/70503924/236116495-aae4474b-dd0c-4a19-b451-ca0429442a43.png"> I tried to check the logs and all of the errors are like this: <img alt="image" width="962" src="https://user-images.githubusercontent.com/70503924/236116626-86c2efea-8f3a-4be2-a85e-19dddd596551.png">
> 
> I still have no idea what caused the error, can anybody help me?
> 
> ### Additional information
> My wsgi.py is like this: <img alt="image" width="750" src="https://user-images.githubusercontent.com/70503924/236116750-8a3e5bb0-2ca7-42a9-bf9d-1e22f6ff5df7.png"> And in the error logs it says Unknown application error occurred Runtime.Unknown
> 
> ### Example
> _No response_

Apparently you cannot deploy django with models to vercel unless you use railway free database hosting. So as long as your using django models you cannot deploy to vercel without railway. Try PythonAywhere.com instead the only difficulty there is retrieving uploaded files on their server but their help forum is awesome!!!   here is speaking ChatGPT, cartoon yourself, job scraper django site : http://mungruez.pythonAnywhere.com

**Comment by azihassan** [comment]
I ran into this issue and in my case, it turns out I was deploying Django 3.x and the official example had Django 4.x. An upgrade in requirements.txt fixed the runtime unknown issue. I still had to fix env vars because they got added with extra line breaks for some reason, so it was giving a 400 because the ALLOWED_HOSTS variable had '.vercel.app\n' instead of '.vercel.app'. But other than that it was smooth sailing.

Here's the official example for reference : https://github.com/vercel/examples/tree/main/python/django

**Reply by mungruez** [reply]
I upgraded to latest version of Django 4.2 and getting the same errors. I am sure this is the case with other users ?

**Reply by azihassan** [reply]
Oh OK, my bad. Just saw in your screenshot that you were already on Django 4 :

![image](https://github.com/vercel/community/assets/1638227/1207dab0-82c5-43cb-82c4-9387f2cb3025)

Might be a long shot, but if it's not already the case, in settings.py, set WSGI_APPLICATION to `propensid02.wsgi.app` instead of `propensid02.wsgi.application`

**Comment by huncholane** [comment]
![kekw](https://github.com/vercel/community/assets/113503551/aff4eb64-1ede-4a39-8495-513e2b808226)

**Comment by lmjaedentai** [comment]
same here :(

**Comment by Ev357** [comment]
i had a similar issue actually, i found this vercel docs: https://vercel.com/templates/python/django-hello-world

i renamed my application to `app` in `project/asgi.py`,
and my vercel.json is looking like this:
```
{
  "builds": [
    {
      "src": "project/asgi.py",
      "use": "@vercel/python"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "project/asgi.py"
    }
  ]
}
```

feel free to replace `asgi` with `wsgi` if needed.

**Reply by samuelh2005** [reply]
Thanks! This worked for me.

**Reply by mungruez** [reply]
Thanks. Hope this solves the issues with deploying Django to vercel.

**Comment by VishalYadavGit** [comment]
Guys just use app=application in wsgi.py at the last line.
this is happening because the vercel is trying to recognize app but there is "application" in wsgi.py instead of "app" so just do this and your project will be deployed

![image](https://github.com/vercel/community/assets/114361635/3bbb43c0-3541-464b-8936-07b56755e24a)

**Reply by samuelh2005** [reply]
Its recommend to use code blocks instead of images. It makes it easier to copy the code. You can do so like
~~~
```python
x = 1
```
~~~

**Reply by VishalYadavGit** [reply]
I am sorry It was my first reply, I will be careful in future. Thanks for the feedback.

**Comment by CoderMungan** [comment]
Hi guys,

Actually i was a same issue! 

I'm sure most of you have tried everything mentioned in this issue and it didn't work. But the real problem isn't here. It stems from incompatible packages in the requirements.txt file. Please check the packages that are compatible with Vercel and try again. I was getting this error because of the typing package. Once I removed it, I was able to deploy very easily.

Happy coding to you all.

And also if u interesting my config;

`vercel.json` file
```json
{
  "version": 2,
  "builds": [
    {
      "src": "backend/wsgi.py",
      "use": "@vercel/python",
      "config": { "maxLambdaSize": "15mb", "runtime": "python3.9" }
    },
    {
      "src": "build_files.sh",
      "use": "@vercel/static-build",
      "config": {
        "distDir": "staticfiles_build"
      }
    }
  ],
  "routes": [
    {
      "src": "/static/(.*)",
      "dest": "/static/$1"
    },
    {
      "src": "/(.*)",
      "dest": "backend/wsgi.py"
    }
  ]
}
```
`wsgi.py`

```python
import os

from django.core.wsgi import get_wsgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'backend.settings')

application = get_wsgi_application()

app = application
```

`settings.py`

```python

ALLOWED_HOSTS = [".vercel.app", ".now.sh"]
...
WSGI_APPLICATION = 'backend.wsgi.application'
...
STATIC_URL = "static/"
STATICFILES_DIRS = os.path.join(BASE_DIR, "static"),
STATIC_ROOT = os.path.join(BASE_DIR, "staticfiles_build","static")
```

**Reply by rowan3373** [reply]
what kind of incompatible package did you have??

**Reply by CoderMungan** [reply]
psycopg2 instead of using this use this psycopg2-binary

**Reply by CoderMungan** [reply]
also which package i used: 

```bash
asgiref==3.7.2
boto3==1.34.34
botocore==1.34.34
Django==4.2.7
django-autoslug==1.9.9
django-cors-headers==4.3.1
django-jazzmin==2.6.0
django-js-asset==2.2.0
django-phonenumber-field==7.3.0
django-storages==1.14.2
django-tinymce==3.6.1
djangorestframework==3.14.0
drf-yasg==1.21.7
inflection==0.5.1
jmespath==1.0.1
packaging==23.2
phonenumberslite==8.13.28
pillow==10.2.0
psycopg2-binary==2.9.9
python-dateutil==2.8.2
python-dotenv==1.0.1
pytz==2023.3.post1
PyYAML==6.0.1
s3transfer==0.10.0
six==1.16.0
sqlparse==0.4.4
tzdata==2023.4
uritemplate==4.1.1
```

**Reply by donkruger** [reply]
This worked for me. Thanks. I was simply missing `psycopg2-binary` in requirements.txt

**Comment by Viral-Sachde** [comment]
same issue here

**Comment by Douglasemmanuel** [comment]
Please could anyone help me out , I got 500 servers less function crashed after deploying Django on vercel

**Reply by VishalYadavGit** [reply]
> Please could anyone help me out , I got 500 servers less function crashed after deploying Django on vercel 
> 

Please share your wsgi.py and vercel.json for reference

**Reply by Douglasemmanuel** [reply]
My wsgi.py


import os
import sys
from django.core.wsgi import get_wsgi_application


os.environ.setdefault("DJANGO_SETTINGS_MODULE", "verceldeploy.settings")

application = get_wsgi_application()
app = application


my VERCEL.JSON
{ 
    "builds": [ 
      { 
        "src": "verceldeploy/wsgi.py", 
        "use": "@vercel/python" 
      } 
    ], 
    "routes": [ 
      { 
        "src": "/(.*)", 
        "dest": "verceldeploy/wsgi.py" 
      } 
    ] 
}

**Comment by dooropening** [comment]
same issue, but i use python + flask, project runs ok on my local machine. it seems "get/favicon" had some problem,it showed that:LAMBDA_WARNING: Unhandled exception. The most likely cause is an issue in the function code. However, in rare cases, a Lambda runtime update can cause unexpected function behavior. For functions using managed runtimes, runtime updates can be triggered by a function change, or can be applied automatically. To determine if the runtime has been updated, check the runtime version in the INIT_START log entry. If this error correlates with a change in the runtime version, you may be able to mitigate this error by temporarily rolling back to the previous runtime version. For more information, see https://docs.aws.amazon.com/lambda/latest/dg/runtimes-update.html
[ERROR] Runtime.ImportModuleError: Unable to import module 'vc__handler__python': No module named 'attribute'.

**Comment by kssvinay** [comment]
**change** the nodejs version **from 20.x to 18.x** in settings of your project

setting > general > nodejs version

[![Nodejs Version][1]][1]


  [1]: https://i.sstatic.net/cWH4ZFPg.png

**Comment by Chibike-267** [comment]
I got thesame error but i'm using nestjs
it displays this error on the browser 404: NOT_FOUND Code: NOT_FOUND ID: cdg1::74czp-1717258485890-4b2b997d9a08; then when tested on my postman or thunder client it displays one html error i don't know about

**Comment by Musheer0** [comment]
does vercel support pytube ? or should i remove it?
